### ISSUE TITLE
automatic opportunistic key verification

### ISSUE THREAD

**Issue by testbird** [autocrypt/autocrypt#394]
It is a very basic  feature that users can manually enter (type) an address to contact somebody.

But nowadays contacts can also be easily imported by scanning QR codes. And the scanned codes may also contain the public key data. Based on this, I think it should be possible to devise an  autocrypt spec level increment that introduces automatic "opportunistic key verification" into usual QR code contact exchanges.

It may be possible to do this without the user having to learn anything about key verfication. Any new UI and UX terminology would only be explicative information for the user. All the while the email application manages the users personal "autocrypt web of trust", without them having to think about it, because it's all build on simple and commonly understandable actions, principles and wording.

---

### Opportunistic Key Verification

Oportunistic key verfication doesn't require full key verification, but makes it as easy as possible by incorporating it into the usual QR code based contact exchange workflow. The intention is to automatically make the regular usage as trustworthy as possible, providing the interested user with short and understandable information about the actual state of affairs, but of course also to allow for intentional key verification.

The basic theme is: Contact details are made easy to pass on privately to other users through QR code displaying and scanning (a separate out-of-network-band channel). (The hardware requirement for a 2-way scanning activity (that supports both displaying and scanning QR codes simultaneously at the same time) is already sufficiently common in several device classes.)

All the users would have to do is, to scan each others qr codes. Everything else can happen automatically.

* Scanning a QR from somebody (that contains the address and public key) automatically adds a `key scanned` label to the contact. (Assumes the user to scan from the correct and trust-able person, i.e. no further requirements for checking codes or similar.)

* Some optional steps like the user entering a user-generated? string and also telling this checking code to the other party for inclusion (signed) in the displayed qr, could also warrant a `key verfied` label.

* Receiving a message from a `key scanned` contact that is also encrypted for another recipient and "gossips" the public key of that other contact together with a (detached) signature for the key, automatically adds a `key referred-by` label to that email-approved "known contact" (e.g. using links in the local contacts database indicating that the contact B got a referred-by status from contact A, and possibly others).

When adding members to groups:

* Groups automatically get a `referral group` trust level label if there are no members without a `key referred-by`, `key scanned,` or `key verified` label.

* Groups automatically get the `key scanned` or `keys verfied` label if all members have the respective label.

* Groups may optionally be created or set to require all members to have at least the `key referred-by`, `key scanned`, or `key verified` status.

The QR scanning activity of autocrypt email applications should also make it really easy for the user to manually pass on other contacts and make **QR code referrals**. Entering the simultaneous QR code displaying and scanning activity should always be only one click away, and also be an option when creating new contacts or groups, or browsing existing contacts.

* Entering the applications default 2-way scanning QR code activity (i.e. not coming from particular contacts' details context) should by default show the users own QR code contact data (at least address & key). It should allow to subsequently scan multiple QR codes (addresses and keys) of others, and add them to the list of contacts with the appropriate labels.

* The default "show own QR code" mode should also have a button to "share another contact", which opens the address book selection, to display a *referral QR code* for the other contact, signed by the local user.

* Scanning a QR code while displaying the signed referral QR code for another contact should correspondingly send a referral email message to the displayed contact (gossiping the public key of the scanned contact, together with a detached signature for it). This allows the displayed contact (that may be passed on by others scanning) to get the contact data of any scanned contacts (and verify the signature to grant the `key referred-by` list or link entry), even if not locally present.

By selecting a group from the contact list (for example in the QR scanning activity) it is also possible to add a contact to a group by scanning its QR code. The key of the newly scanned contact is then sent to all group members (including the newly scanned contact) within the referral email that adds the new member and gossips all keys of the members and their detached signatures by the referrer.

**Comment by Valodim** [comment]
This is a bit out of scope, still. We want to get key discovery and robust automated public key management working before we start a trust model discussion, and this seems to dive right into that :)
